### PR TITLE
Mark operands of dead FIELD_LIST as unused

### PR DESCRIPTION
**Pull request by CarolEidt**

### CODE REVIEW

**Comment by CarolEidt** [comment]
This assert, and the underlying issue, was encountered by two different ILGEN tests - one that triggered the issue on arm32, and one that triggered it on x86. I've extracted the relevant portion of both tests in the test case.

**Comment by CarolEidt** [comment]
@dotnet/jit-contrib PTAL

File: src/jit/lir.h [hunk 1]
```diff
@@ -314,6 +314,20 @@ class LIR final
...
+    {
+        GenTreeFieldList* listNode = this->AsFieldList();
+        while (listNode != nullptr)
+        {
+            listNode->Current()->SetUnusedValue();
+            GenTreeFieldList* nextListNode = listNode->Rest();
+            listNode->gtBashToNOP();
+            listNode = nextListNode;
+        }
+        return;
```
Comment (sandreenko) [inline src/jit/lir.h]: We do not set `UnusedValue` for the list itself, is it correct?

**Comment by CarolEidt** [comment]
> I do not understand what is the difference between setting GT_FIELD_LIST as unused, and for example GT_ADD, in the case of GT_ADD are children marked as unused?

The difference is due to the weirdness of `GT_FIELD_LIST`, in that only the head of the list is in the execution list (which, IMO, was probably a bad design decision), but the child of each list node *is* in the execution list. So they won't be encountered if the list isn't traversed, and will wind up as unused but not marked accordingly.

> We do not set UnusedValue for the list itself, is it correct?

A `GT_FIELD_LIST` can't be marked as unused, because it's required to be contained. So instead we bash it to a `NOP`.

**Comment by sandreenko** [comment]
> A GT_FIELD_LIST can't be marked as unused, because it's required to be contained. So instead we bash it to a NOP.

Code after does:
```
#ifndef LEGACY_BACKEND
     ClearContained();
 #endif
```
and you have already bashed it to nop,
so the nop node is not required to be contained,
this code clear contained flag, why can't we then set unused to the NOP node that is not contained?

**Comment by CarolEidt** [comment]
@sandreenko I'm not sure I understand the question. The bashed NOP is itself eliminated in the backward traversal in dead code elimination. We could mark it as unused and clear containment, but that's necessary for a NOP.

**Comment by CarolEidt** [comment]
@dotnet-bot retest Windows_NT arm Cross Checked Innerloop Build and Test
@dotnet-bot retest Windows_NT arm64 Cross Checked Innerloop Build and Test
@dotnet-bot retest Windows_NT armlb Cross Checked Innerloop Build and Test

**Comment by CarolEidt** [comment]
@dotnet-bot retest Windows_NT arm Cross Checked Innerloop Build and Test

**Comment by CarolEidt** [comment]
@dotnet-bot retest Windows_NT arm Cross Checked Innerloop Build and Test

**Comment by sandreenko** [comment]
@CarolEidt , I made a branch on top of your changes, could you please take a look? 
Fix is [there ](https://github.com/sandreenko/coreclr/commit/0cd3c945872fb31a1e824526e5321e1d1115345f)
and some additional printing is [there ](https://github.com/sandreenko/coreclr/commit/51c7f7896291c33307f21732b5c17d754f92f8fc).

The idea is that it is not `SetUnusedValue`'s responsibility to handle this case, it should be a simple function that only set flags. Also `SetUnusedValue` is called from other places too and some of them do not except such behavior, so I think we should fix `GT_FIELD_LIST` handling in `fgComputeLifeLIR` (I think this case was just missed there when another hack for `GT_PUTARG_STK` was added:
```
                        // Special-case PUTARG_STK: since this operator is not considered a value, DCE will not remove
                        // these nodes.
                        else if (operand->OperIs(GT_PUTARG_STK))
                        {
                            assert(!operand->IsValue());
                            GenTreePutArgStk* putArg = operand->AsPutArgStk();
                            putArg->gtOp1->SetUnusedValue();
                            JITDUMP("Bash dead put_arg_stk to nop:\n");
                            DISPNODE(putArg);
                            putArg->gtBashToNOP();
                        }
```
and this hack is the only place where we get unused `GT_FIELD_LIST`.

**Comment by CarolEidt** [comment]
> and this hack is the only place where we get unused GT_FIELD_LIST.

I disagree that this is the right place to handle this. It may be the case for now that a `GT_FIELD_LIST` can only become unused in the case of eliminating a dead call, but that doesn't ensure that it won't set unused on a `GT_FIELD_LIST` in some other context in future. The alternative would be messy handling in `fgComputeLife()` *plus* an assert in `SetUnusedValue` because this is the one value node that can't be marked unused.

**Comment by CarolEidt** [comment]
@sandreenko - thinking about this further, I'm revising my view.
Fundamentally, this issue is caused by the fact that 1) `GT_FIELD_LIST` has "effective" children that are not actual children. Perhaps an approach that fixed the traversal to traverse the entire list (as is done for other nodes that have lists as children - this one is just extra weird because its children have the same operator). I'll have a further look.

**Comment by sandreenko** [comment]
My idea is that we should handle unused GT_FIELDS right, so our design says that only the head is represented in execution order - ok, so when we see a GT_FIELDS head that is marked as unused - we should take care of its children.
In other workds `tree->SetUnusedValue(); tree->ClearUnusedValue();` should not produce diffs.

**Comment by CarolEidt** [comment]
I made the necessary changes to:

- Mark the children of an unused `GT_FIELD_LIST` as unused. The change needed for this to work is to not look at the GTF_CALL flag on the node. The nodes in that case are not calls, and in that case it's not necessary to worry about the children (any actual calls with side-effects won't be removed).
- Gracefully handle the case where, for some reason, a `GT_FIELD_LIST` is marked unused, but not deleted. I felt this was necessary for safety. I tested the fix without the first change, to make sure it works, though arguably this defensive code will not get ongoing testing.

File: src/jit/liveness.cpp [hunk 2]
```diff
@@ -2225,7 +2225,8 @@ void Compiler::fgComputeLifeLIR(VARSET_TP& life, BasicBlock* block, VARSET_VALAR
                 assert(!node->OperIsLocal());
                 if (!node->IsValue() || node->IsUnusedValue())
                 {
-                    unsigned sideEffects = node->gtFlags & (GTF_SIDE_EFFECT | GTF_SET_FLAGS);
+                    // We are not interested in GTF_CALL, which will be set on nodes with calls in their subtree.
+                    unsigned sideEffects = node->gtFlags & (GTF_ASG | GTF_EXCEPT | GTF_SET_FLAGS);
```
Comment (sandreenko) [inline src/jit/liveness.cpp]: For `GTF_EXCEPT` we check that `((sideEffects == GTF_EXCEPT) && !node->OperMayThrow(this))` so it means `remove if the node itself doesn't throw` (We have a checker that guarantee that there are no nodes without `GTF_EXCEPT` which `OperMayThrow`).

but my question is do we care about `GTF_ASG` from subtrees?

**Comment by CarolEidt** [comment]
> How does dce results look like now? Does it set unused and delete only the head node of unused field lists? Does it delete field_list's children?

For the test case, we remove the dead children with the liveness change (not without). I haven't yet run full diffs.

File: src/jit/liveness.cpp (hunk 2 above)
Comment (CarolEidt) [reply src/jit/liveness.cpp]: We probably don't care about `GTF_ASG` from subtrees. What we probably need here is the equivalent of `Compiler::gtUpdateNodeOperSideEffects()` that only returns the side effects of the current node, e.g. `Compiler::getNodeOperSideEffects()`

File: src/jit/liveness.cpp [hunk 3]
```diff
@@ -2225,8 +2225,11 @@ void Compiler::fgComputeLifeLIR(VARSET_TP& life, BasicBlock* block, VARSET_VALAR
                 assert(!node->OperIsLocal());
                 if (!node->IsValue() || node->IsUnusedValue())
                 {
-                    // We are not interested in GTF_CALL, which will be set on nodes with calls in their subtree.
-                    unsigned sideEffects = node->gtFlags & (GTF_ASG | GTF_EXCEPT | GTF_SET_FLAGS);
+                    // We are only interested in avoiding the removal of nodes with direct side-effects
+                    // (as opposed to side effects of their children).
+                    // This default case should never include calls or assignments.
+                    assert(!node->OperRequiresAsgFlag() && !node->OperIs(GT_CALL));
+                    unsigned sideEffects = node->gtFlags & (GTF_EXCEPT | GTF_SET_FLAGS);
```
Comment (sandreenko) [inline src/jit/liveness.cpp]:
```
unsigned sideEffects = node->gtFlags & (GTF_EXCEPT | GTF_SET_FLAGS); 
if ((sideEffects == 0) || ((sideEffects == GTF_EXCEPT) && !node->OperMayThrow(this))) 
```
equal to 
```
if (!node->gtSetFlags() && !node->OperMayThrow(this)) 
```

```

**Comment by sandreenko** [comment]
> Gracefully handle the case where, for some reason, a GT_FIELD_LIST is marked unused, but not deleted. I felt this was necessary for safety. I tested the fix without the first change, to make sure it works, though arguably this defensive code will not get ongoing testing.

Do we have such cases now? Are they bugs that need to be fixed/ need an issue to fix it later and delete this new code?

**Comment by CarolEidt** [comment]
> Do we have such cases now? Are they bugs that need to be fixed/ need an issue to fix it later and delete this new code?

We don't have such a case, but I don't know of a good way to defensively protect against setting a `GT_FIELD_LIST` as unused, and having it blow up downstream.

BTW - the x86 diffs finished, and there is exactly one case where the relaxing of the side-effect condition caused the removal of a dead `PUTARG_REG` node, resulting in a 2 byte reduction.

**Comment by sandreenko** [comment]
> BTW - the x86 diffs finished, and there is exactly one case where the relaxing of the side-effect condition caused the removal of a dead PUTARG_REG node, resulting in a 2 byte reduction.
Good.

changes in liveness.cpp LGTM, I think it is the right fix, only the head is marked as unused and only the head is actually removed from the execution order.

However, I think changes in other files are harmful, we should not add code that is unreachable now and that we can't test (and with commented asserts that won't fire). 

> but that doesn't ensure that it won't set unused on a GT_FIELD_LIST in some other context in future. 

Could we create a separate issue to decide how to handle or prevent these future cases? it is differently not a part of this issue fix. So if somebody add such changes he will better see asserts rather then silent bad codegeneration for example.

**Comment by CarolEidt** [comment]
> Could we create a separate issue to decide how to handle or prevent these future cases? it is differently not a part of this issue fix. So if somebody add such changes he will better see asserts rather then silent bad codegeneration for example.

The problem is that an end-user will never see an assert - they will see SBCG anyway because there is no existing code to handle that case.
This is my fundamental concern with this approach to the bug fix. This bug was hit with a relatively obscure case (it was only discovered with ilgen), and so it is possible that another caller of `SetUnusedValue` could mark a `GT_FIELD_LIST` as unused, and we would simply generate bad code in Release. There is no correct way to handle that situation without either a `noway_assert` or the code I've added.

**Comment by sandreenko** [comment]
Thanks for your the explanation. 
Last changes LGTM, thanks.

**Pull request merged**